File: supabase/functions/_shared/summary.ts
/**
 * The figures the Mac widget shows.
 *
 * Kept apart from the endpoint so it can be tested without a database: a
 * widget that quietly renders a stale or wrong number is worse than one that
 * fails visibly, and there is nowhere in a widget to surface a mistake.
 */

export interface CycleRow {
  day: string
  strain: number | null
  avg_hr: number | null
  max_hr: number | null
  kilojoules: number | null
}

export interface RecoveryRow {
  day: string
  recovery_pct: number | null
  hrv_ms: number | null
  resting_hr: number | null
}

export interface SleepRow {
  day: string
  asleep_min: number | null
  need_min: number | null
  performance_pct: number | null
  efficiency_pct: number | null
}

export type ReadinessBand = 'recover' | 'pace' | 'ready' | 'go'

export interface Summary {
  day: string | null
  recovery: number | null
  hrv: number | null
  restingHr: number | null
  strain: number | null
  calories: number | null
  /** The cycle's average and peak heart rate, in bpm. */
  avgHr: number | null
  maxHr: number | null
  sleepMin: number | null
  sleepNeedMin: number | null
  sleepPerformance: number | null
  sleepEfficiency: number | null
  /** 0-10, the same blend the dashboard's Insights tab shows. Null until
   *  there is at least one of recovery/sleep/HRV/resting heart rate to
   *  build it from. */
  readiness: number | null
  readinessBand: ReadinessBand | null
  /** Oldest to newest, one entry per day that has a reading. */
  recoveryTrend: Array<{ day: string; value: number }>
  strainTrend: Array<{ day: string; value: number }>
  hrvTrend: Array<{ day: string; value: number }>
  /** Minutes asleep per day, for the large widget's sleep trend. */
  sleepTrend: Array<{ day: string; value: number }>
  restingHrTrend: Array<{ day: string; value: number }>
  updatedAt: string
}

const byDay = <T extends { day: string }>(rows: T[]): Map<string, T> => {
  const m = new Map<string, T>()
  for (const r of rows) if (r.day) m.set(r.day, r)
  return m
}

const trend = <T extends { day: string }>(
  rows: T[], pick: (r: T) => number | null,
): Array<{ day: string; value: number }> =>
  rows
    .filter((r) => r.day)
    .map((r) => ({ day: r.day, value: pick(r) }))
    .filter((p): p is { day: string; value: number } => p.value !== null)
    .sort((a, b) => a.day.localeCompare(b.day))

const mean = (xs: number[]): number | null =>
  xs.length ? xs.reduce((s, x) => s + x, 0) / xs.length : null

/** Scale a value to 0..1 across a plausible range, clamped at both ends. */
const norm = (v: number, lo: number, hi: number): number =>
  Math.max(0, Math.min(1, (v - lo) / (hi - lo)))

/** How far above or below a baseline, expressed as 0..1 where 0.5 is "on
 *  baseline". `spread` is the deviation that maps to the top or bottom. */
const relative = (v: number, baseline: number, spread: number, higherIsBetter: boolean): number => {
  const delta = (v - baseline) / (spread || 1)
  const scaled = 0.5 + (higherIsBetter ? delta : -delta) * 0.5
  return Math.max(0, Math.min(1, scaled))
}

const bandFor = (score: number): ReadinessBand => {
  if (score < 4) return 'recover'
  if (score < 6) return 'pace'
  if (score < 8) return 'ready'
  return 'go'
}

/**
 * The same 0-10 readiness blend as `computeReadiness` in the dashboard's
 * `src/lib/wellness.ts` -- recovery, sleep, HRV and resting heart rate
 * weighted 0.4/0.3/0.2/0.1, with HRV and resting heart rate judged against
 * their own trailing baseline rather than a fixed scale. Reimplemented here
 * rather than shared, since this edge function already has exactly the rows
 * it needs (the same `DAYS`-day window fetched for the trend arrays) and
 * pulling in the dashboard's richer client-side types would be more coupling
 * than the duplication it would save.
 */
function computeReadiness(
  day: string,
  recovery: RecoveryRow[],
  sleep: SleepRow[],
  priorDays: string[],
): { score: number | null; band: ReadinessBand | null } {
  const recByDay = byDay(recovery)
  const sleepByDay = byDay(sleep)
  const rec = recByDay.get(day)
  const sl = sleepByDay.get(day)

  const priorHrv = priorDays.map((d) => recByDay.get(d)?.hrv_ms).filter((v): v is number => v != null)
  const priorRhr = priorDays.map((d) => recByDay.get(d)?.resting_hr).filter((v): v is number => v != null)

  const parts: Array<{ weight: number; value: number | null }> = []

  parts.push({
    weight: 0.4,
    value: rec?.recovery_pct != null ? norm(rec.recovery_pct, 0, 100) : null,
  })

  parts.push({
    weight: 0.3,
    value: sl?.performance_pct != null ? norm(sl.performance_pct, 40, 100)
      : sl?.asleep_min != null ? norm(sl.asleep_min, 240, 510)
      : null,
  })

  const hrvBase = mean(priorHrv)
  const hrvSpread = priorHrv.length > 4 ? (Math.max(...priorHrv) - Math.min(...priorHrv)) / 2 : null
  parts.push({
    weight: 0.2,
    value: rec?.hrv_ms != null && hrvBase != null && hrvSpread
      ? relative(rec.hrv_ms, hrvBase, hrvSpread, true) : null,
  })

  const rhrBase = mean(priorRhr)
  const rhrSpread = priorRhr.length > 4 ? (Math.max(...priorRhr) - Math.min(...priorRhr)) / 2 : null
  parts.push({
    weight: 0.1,
    value: rec?.resting_hr != null && rhrBase != null && rhrSpread
      ? relative(rec.resting_hr, rhrBase, rhrSpread, false) : null,
  })

  const present = parts.filter((p) => p.value !== null)
  if (!present.length) return { score: null, band: null }

  const totalWeight = present.reduce((s, p) => s + p.weight, 0)
  const blended = present.reduce((s, p) => s + p.value! * p.weight, 0) / totalWeight
  const score = Math.round(blended * 100) / 10
  return { score, band: bandFor(score) }
}

/**
 * The most recent day with anything on it.
 *
 * Not simply "today": Whoop scores a night when you wake, so on an overnight
 * shift the newest complete day is often yesterday's date, and a widget that
 * asked for today alone would show blanks for most of the morning.
 */
export function summarise(
  cycles: CycleRow[],
  recovery: RecoveryRow[],
  sleep: SleepRow[],
  now: Date = new Date(),
): Summary {
  const recoveryTrend = trend(recovery, (r) => r.recovery_pct)
  const strainTrend = trend(cycles, (c) => c.strain)
  const hrvTrend = trend(recovery, (r) => r.hrv_ms)
  const sleepTrend = trend(sleep, (row) => row.asleep_min)
  const restingHrTrend = trend(recovery, (r) => r.resting_hr)

  const days = [...cycles, ...recovery, ...sleep].map((r) => r.day).filter(Boolean).sort()
  const day = days.length ? days[days.length - 1] : null

  const c = day ? byDay(cycles).get(day) : undefined
  const r = day ? byDay(recovery).get(day) : undefined
  const s = day ? byDay(sleep).get(day) : undefined

  // Every day strictly before the latest one, for the readiness baseline.
  const priorDays = day ? days.filter((d) => d < day) : []
  const readiness = day
    ? computeReadiness(day, recovery, sleep, priorDays)
    : { score: null, band: null }

  return {
    day,
    recovery: r?.recovery_pct ?? null,
    hrv: r?.hrv_ms ?? null,
    restingHr: r?.resting_hr ?? null,
    strain: c?.strain ?? null,
    calories: c?.kilojoules != null ? Math.round(c.kilojoules / 4.184) : null,
    avgHr: c?.avg_hr ?? null,
    maxHr: c?.max_hr ?? null,
    sleepMin: s?.asleep_min ?? null,
    sleepNeedMin: s?.need_min ?? null,
    sleepPerformance: s?.performance_pct ?? null,
    sleepEfficiency: s?.efficiency_pct ?? null,
    readiness: readiness.score,
    readinessBand: readiness.band,
    recoveryTrend,
    strainTrend,
    hrvTrend,
    sleepTrend,
    restingHrTrend,
    updatedAt: now.toISOString(),
  }
}

File: mac-widget/Shared/Summary.swift
import Foundation
import SwiftUI

/// The payload `whoop-widget` returns. The field names match the JSON exactly,
/// so no key mapping is needed; anything Whoop has not scored yet arrives null
/// and stays optional all the way to the view, which draws a dash for it.
struct TrendPoint: Decodable, Hashable {
    let day: String
    let value: Double
}

struct Summary: Decodable {
    let day: String?
    let recovery: Double?
    let hrv: Double?
    let restingHr: Double?
    let strain: Double?
    let calories: Double?
    /// The cycle's average and peak heart rate. Optional in the model as well
    /// as in the data: a widget refreshing before the backend redeploys would
    /// otherwise fail to decode the whole payload over one missing key.
    let avgHr: Double?
    let maxHr: Double?
    let sleepMin: Double?
    let sleepNeedMin: Double?
    let sleepPerformance: Double?
    let sleepEfficiency: Double?
    /// 0…10, the same blend the dashboard's Insights tab shows.
    let readiness: Double?
    /// "recover" | "pace" | "ready" | "go", or nil alongside a nil `readiness`.
    let readinessBand: String?
    let recoveryTrend: [TrendPoint]
    let strainTrend: [TrendPoint]
    /// Optional, unlike the other two trends, for the same reason as `avgHr`:
    /// added after the first widget shipped, and a missing key must degrade to
    /// a dash rather than fail the whole decode.
    let hrvTrend: [TrendPoint]?
    let sleepTrend: [TrendPoint]?
    let restingHrTrend: [TrendPoint]?
    let updatedAt: String
}

/// How a metric's latest value compares to the rest of its trend window,
/// mirroring the dashboard's "vs 28-day average" callouts at the widget's
/// much shorter horizon.
struct TrendDelta {
    enum Direction { case up, down, flat }
    let direction: Direction
    let magnitudeText: String

    var symbol: String {
        switch direction {
        case .up: return "↑"
        case .down: return "↓"
        case .flat: return "→"
        }
    }
}

/// `format` is for a metric whose raw unit does not read as a number beside
/// its own value: sleep is held in minutes, so a bare "112" next to "5h 20m"
/// says nothing until it is rendered as a duration too.
private func trendDelta(
    _ points: [TrendPoint],
    decimals: Int,
    format: ((Double) -> String)? = nil
) -> TrendDelta? {
    guard points.count > 1, let latest = points.last?.value else { return nil }
    let prior = points.dropLast().map(\.value)
    let mean = prior.reduce(0, +) / Double(prior.count)
    let diff = latest - mean
    let direction: TrendDelta.Direction = abs(diff) < 0.05 ? .flat : (diff > 0 ? .up : .down)
    let magnitude = format?(abs(diff)) ?? String(format: "%.\(decimals)f", abs(diff))
    return TrendDelta(direction: direction, magnitudeText: magnitude)
}

/// A trend window reduced to what the large layout labels a sparkline with:
/// how many days it covers, the average across them, and the range the line
/// spans. A sparkline on its own shows a shape without a scale; these are the
/// numbers that give it one.
struct TrendStats {
    let count: Int
    let average: Double
    let low: Double
    let high: Double
    let latest: Double

    /// `days` takes the most recent n points rather than the whole window,
    /// since the payload carries thirty days and the large layout quotes a
    /// week. Returns nil for an empty trend, so the caller draws a dash
    /// instead of a zero that looks like a reading.
    init?(_ points: [TrendPoint], days: Int? = nil) {
        let window = days.map { Array(points.suffix($0)) } ?? points
        let values = window.map(\.value)
        guard let latest = values.last,
              let low = values.min(),
              let high = values.max() else { return nil }
        self.count = values.count
        self.average = values.reduce(0, +) / Double(values.count)
        self.low = low
        self.high = high
        self.latest = latest
    }

    func averageText(decimals: Int = 0, unit: String = "") -> String {
        let number = String(format: "%.\(decimals)f", average)
        return unit.isEmpty ? number : "\(number)\(unit)"
    }

    func rangeText(decimals: Int = 0, unit: String = "") -> String {
        let lowText = String(format: "%.\(decimals)f", low)
        let highText = String(format: "%.\(decimals)f", high)
        return unit.isEmpty ? "\(lowText)–\(highText)" : "\(lowText)–\(highText)\(unit)"
    }
}

/// "8h 32m", the one duration format the whole widget uses.
func durationText(_ minutes: Double) -> String {
    let total = Int(minutes.rounded())
    let hours = total / 60
    let rest = total % 60
    return hours > 0 ? "\(hours)h \(rest)m" : "\(rest)m"
}

enum SummaryError: LocalizedError {
    case notConfigured
    case badURL
    case http(Int)

    var errorDescription: String? {
        switch self {
        case .notConfigured: return "Add your token in Config.swift"
        case .badURL: return "The endpoint in Config.swift is not a URL"
        case .http(401): return "Token rejected"
        case .http(let code): return "Server said \(code)"
        }
    }
}

func fetchSummary() async throws -> Summary {
    guard !Config.token.isEmpty, !Config.token.hasPrefix("PASTE_") else {
        throw SummaryError.notConfigured
    }
    guard let url = URL(string: Config.endpoint) else { throw SummaryError.badURL }

    var request = URLRequest(url: url)
    request.setValue("Bearer \(Config.token)", forHTTPHeaderField: "Authorization")
    // A widget that redraws a cached body has no way to show it is stale, so
    // every refresh goes to the network.
    request.cachePolicy = .reloadIgnoringLocalCacheData
    request.timeoutInterval = 20

    let (data, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse else { throw SummaryError.http(0) }
    guard http.statusCode == 200 else { throw SummaryError.http(http.statusCode) }
    return try JSONDecoder().decode(Summary.self, from: data)
}

// MARK: - Display

extension Summary {
    var recoveryText: String { recovery.map { "\(Int($0.rounded()))%" } ?? "—" }
    var strainText: String { strain.map { String(format: "%.1f", $0) } ?? "—" }
    var hrvText: String { hrv.map { "\(Int($0.rounded())) ms" } ?? "—" }
    var restingHrText: String { restingHr.map { "\(Int($0.rounded())) bpm" } ?? "—" }
    var caloriesText: String { calories.map { "\(Int($0.rounded())) kcal" } ?? "—" }
    var readinessText: String { readiness.map { String(format: "%.1f", $0) } ?? "—" }
    var avgHrText: String { avgHr.map { "\(Int($0.rounded())) bpm" } ?? "—" }
    var maxHrText: String { maxHr.map { "\(Int($0.rounded())) bpm" } ?? "—" }
    var hrvTrendPoints: [TrendPoint] { hrvTrend ?? [] }
    var sleepTrendPoints: [TrendPoint] { sleepTrend ?? [] }
    var restingHrTrendPoints: [TrendPoint] { restingHrTrend ?? [] }

    var sleepText: String {
        guard let minutes = sleepMin else { return "—" }
        return durationText(minutes)
    }

    var sleepNeedText: String { sleepNeedMin.map(durationText) ?? "—" }
    var sleepPerformanceText: String { sleepPerformance.map { "\(Int($0.rounded()))%" } ?? "—" }
    var sleepEfficiencyText: String { sleepEfficiency.map { "\(Int($0.rounded()))%" } ?? "—" }

    /// How much of the night's need was actually slept, 0…1, for the large
    /// layout's meter. Capped at 1: sleeping past the need fills the bar
    /// rather than overflowing it.
    var sleepFraction: Double? {
        guard let sleepMin, let need = sleepNeedMin, need > 0 else { return nil }
        return min(max(sleepMin / need, 0), 1)
    }

    /// "1h 05m short" / "32m over", the gap between slept and needed. Whoop
    /// gives a percentage for the same thing; the duration is what tells you
    /// how much earlier to go to bed.
    var sleepBalanceText: String? {
        guard let sleepMin, let need = sleepNeedMin else { return nil }
        let diff = sleepMin - need
        if abs(diff) < 5 { return "on need" }
        return diff < 0 ? "\(durationText(-diff)) short" : "\(durationText(diff)) over"
    }

    /// Whoop's day strain runs 0…21 on a logarithmic scale, so the fraction is
    /// for a progress meter, not a claim that 10.5 is half a hard day.
    var strainFraction: Double? {
        guard let strain else { return nil }
        return min(max(strain / 21, 0), 1)
    }

    /// 0…1 for the readiness meter, from the same 0…10 score as the badge.
    var readinessFraction: Double? {
        guard let readiness else { return nil }
        return min(max(readiness / 10, 0), 1)
    }

    /// "Recover" / "Pace" / "Ready" / "Go" — the badge's word with room to
    /// breathe, for the large layout.
    var readinessLongLabel: String {
        switch readinessBand {
        case "recover": return "Recover"
        case "pace": return "Pace yourself"
        case "ready": return "Ready"
        case "go": return "Go for it"
        default: return "Readiness"
        }
    }

    /// "Updated 08:42", in the viewer's own timezone and clock format. The
    /// widget cannot say it is stale any other way: a cached body and a fresh
    /// one are drawn identically.
    var updatedAtText: String? {
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        var parsed = iso.date(from: updatedAt)
        if parsed == nil {
            iso.formatOptions = [.withInternetDateTime]
            parsed = iso.date(from: updatedAt)
        }
        guard let date = parsed else { return nil }

        let display = DateFormatter()
        display.dateStyle = .none
        display.timeStyle = .short
        return "Updated \(display.string(from: date))"
    }

    var sleepDetail: String {
        guard let performance = sleepPerformance else { return "Sleep" }
        return "Sleep · \(Int(performance.rounded()))% of need"
    }

    /// "92% perf · 88% eff", trimmed to whichever of the two is present.
    var sleepSecondaryText: String? {
        var parts: [String] = []
        if let performance = sleepPerformance { parts.append("\(Int(performance.rounded()))% perf") }
        if let efficiency = sleepEfficiency { parts.append("\(Int(efficiency.rounded()))% eff") }
        return parts.isEmpty ? nil : parts.joined(separator: " · ")
    }

    /// 0…1, for the ring. Nothing scored yet leaves it empty rather than full.
    var recoveryFraction: Double {
        guard let recovery else { return 0 }
        return min(max(recovery / 100, 0), 1)
    }

    /// Whoop's own bands: green from 67, yellow from 34, red below that.
    var recoveryColor: Color {
        guard let recovery else { return .secondary }
        if recovery >= 67 { return Color(red: 0.06, green: 0.73, blue: 0.51) }
        if recovery >= 34 { return Color(red: 1.00, green: 0.73, blue: 0.09) }
        return Color(red: 0.98, green: 0.31, blue: 0.35)
    }

    /// Recover / Pace / Ready / Go, short enough for the widget's readiness
    /// badge. Falls back to a neutral word if the band is ever unrecognised,
    /// rather than showing nothing.
    var readinessShortLabel: String {
        switch readinessBand {
        case "recover": return "LOW"
        case "pace": return "PACE"
        case "ready": return "READY"
        case "go": return "GO"
        default: return "READY"
        }
    }

    /// Mirrors the dashboard's readiness band colors: red below "pace",
    /// amber for "pace", green from "ready" up.
    var readinessColor: Color {
        switch readinessBand {
        case "recover": return Color(red: 0.98, green: 0.31, blue: 0.35)
        case "pace": return Color(red: 1.00, green: 0.73, blue: 0.09)
        case "ready", "go": return Color(red: 0.06, green: 0.73, blue: 0.51)
        default: return .secondary
        }
    }

    /// Latest recovery vs. the mean of the rest of `recoveryTrend`.
    var recoveryDelta: TrendDelta? { trendDelta(recoveryTrend, decimals: 0) }

    /// Latest strain vs. the mean of the rest of `strainTrend`.
    var strainDelta: TrendDelta? { trendDelta(strainTrend, decimals: 1) }

    /// The same comparison for the metrics only the large layout has room to
    /// show a trend for.
    var hrvDelta: TrendDelta? { trendDelta(hrvTrendPoints, decimals: 0) }
    var restingHrDelta: TrendDelta? { trendDelta(restingHrTrendPoints, decimals: 0) }
    var sleepDelta: TrendDelta? { trendDelta(sleepTrendPoints, decimals: 0, format: durationText) }

    /// A week of each metric, for the large layout's labelled trend rows.
    var recoveryWeek: TrendStats? { TrendStats(recoveryTrend, days: 7) }
    var strainWeek: TrendStats? { TrendStats(strainTrend, days: 7) }
    var hrvWeek: TrendStats? { TrendStats(hrvTrendPoints, days: 7) }
    var restingHrWeek: TrendStats? { TrendStats(restingHrTrendPoints, days: 7) }
    var sleepWeek: TrendStats? { TrendStats(sleepTrendPoints, days: 7) }

    /// How many days the payload's window actually has readings for, so the
    /// large layout can say what its averages are averaging.
    var loggedDays: Int {
        Set(recoveryTrend.map(\.day))
            .union(strainTrend.map(\.day))
            .union(sleepTrendPoints.map(\.day))
            .count
    }

    /// "Mon 14 Sep", or nothing if no day has been scored yet.
    var dayText: String {
        guard let day else { return "No data yet" }
        let parser = DateFormatter()
        parser.dateFormat = "yyyy-MM-dd"
        parser.timeZone = TimeZone(secondsFromGMT: 0)
        guard let date = parser.date(from: day) else { return day }

        let display = DateFormatter()
        display.dateFormat = "EEE d MMM"
        display.timeZone = TimeZone(secondsFromGMT: 0)
        return display.string(from: date)
    }
}
